src/BizHawk.Emulation.Cores/Waterbox/ElfLoader.cs

```diff
@@ -0,0 +1,356 @@
+		public void ConnectSyscalls(IImportResolver syscalls)
+		{
+			Memory.Protect(Memory.Start, Memory.Size, MemoryBlockBase.Protection.RW);
+
+			var tmp = new IntPtr[1];
+
+			foreach (var s in _importSymbols)
+			{
+				if (s.Size == 8)
+				{
+					var p = syscalls.GetProcAddrOrThrow(s.Name);
+					tmp[0] = p;
+					Marshal.Copy(tmp, 0, Z.US(s.Value), 1);
+				}
+				else
+				{
+					if ((s.Size & 7) != 0)
+						throw new InvalidOperationException($"Symbol {s.Name} has unexpected size");
+					var count = (int)(s.Size >> 3);
+					for (var i = 0; i < count; i++)
+					{
+						var p = syscalls.GetProcAddrOrThrow($"{s.Name}[{i}]");
+						tmp[0] = p;
+						Marshal.Copy(tmp, 0, Z.US(s.Value + (ulong)(i * 8)), 1);
+					}
+				}
+			}
+
+			Protect();
+		}
+
+		[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+		private delegate void ElfEntryDelegate();
+
+		public void RunNativeInit()
+		{
+			CallingConventionAdapters.Waterbox.GetDelegateForFunctionPointer<ElfEntryDelegate>(Z.US(_elf.EntryPoint))();
+		}
+
+		public void SealImportsAndTakeXorSnapshot()
+		{
+			if (_everythingSealed)
+				throw new InvalidOperationException($"{nameof(ElfLoader)} already sealed!");
+
+			// save import values, then zero them all (for hash purposes), then take our snapshot, then load them again,
+			// then set the .idata area to read only
+			byte[] impData = null;
+			if (_imports != null)
+			{
+				impData = new byte[_imports.Size];
+				Marshal.Copy(Z.US(_imports.LoadAddress), impData, 0, (int)_imports.Size);
+				WaterboxUtils.ZeroMemory(Z.US(_imports.LoadAddress), (long)_imports.Size);
+			}
+			else
+			{
+				throw new Exception("Call natt (_imports??)");
+			}
+			byte[] invData = null;
+			if (_invisible != null)
+			{
+				invData = new byte[_invisible.Size];
+				Marshal.Copy(Z.US(_invisible.LoadAddress), invData, 0, (int)_invisible.Size);
+				WaterboxUtils.ZeroMemory(Z.US(_invisible.LoadAddress), (long)_invisible.Size);
+			}
+			Memory.SaveXorSnapshot();
+			if (_imports != null)
+			{
+				Marshal.Copy(impData, 0, Z.US(_imports.LoadAddress), (int)_imports.Size);
+			}
+			if (_invisible != null)
+			{
+				Marshal.Copy(invData, 0, Z.US(_invisible.LoadAddress), (int)_invisible.Size);
+			}
+
+			_everythingSealed = true;
+			Protect();
+		}
+
+		private bool _disposed = false;
+
+		public void Dispose()
+		{
+			if (!_disposed)
+			{
+				Memory.Dispose();
+				Memory = null;
+				_disposed = true;
+			}
+		}
+
+		public IntPtr GetProcAddrOrZero(string entryPoint)
+		{
+			if (_visibleSymbols.TryGetValue(entryPoint, out var sym))
+			{
+				return Z.US(sym.Value);
+			}
+			else
+			{
+				return IntPtr.Zero;
+			}
+		}
+
+		public IntPtr GetProcAddrOrThrow(string entryPoint)
+		{
+			if (_visibleSymbols.TryGetValue(entryPoint, out var sym))
+			{
+				return Z.US(sym.Value);
+			}
+			else
+			{
+				throw new InvalidOperationException($"Couldn't find {nameof(entryPoint)} {entryPoint} in {ModuleName}");

```

Comment:
@YoshiRulz Once upon a time, `GetProcAddrOrThrow` was an extension method and not actually in the interface.  It's a lot of noise reimplementing the exact same thing the exact same way every time.  On the other hand, the individual implementations provide more relevant exception messages for each usecase.

Reply:
Extension method sounds like the better option, I'll put it on my to-do list.